### PR TITLE
esc_calibration: increase safety and initialise all data

### PR DESCRIPTION
**Pull request by RomanBapst**
In certain situations it might have been possible to start an ESC calibration even if the battery was still connected. This adds some additional checks to make sure that we have a valid battery measurement.
Furthermore, some data was not initialised and could have possibly led to unexpected behavior.

### CODE REVIEW

**Comment by RomanBapst** [comment]
@dagar Are you suggesting to iterate over multiple instances?

**Comment by dagar** [comment]
I believe that would be the right thing to do, but it looks like commander isn't even doing that yet. https://github.com/PX4/Firmware/blob/master/src/modules/commander/commander.cpp#L1982

At some point we'll need to revisit this and bring it all together (battery_status x N + system_power), because at the moment it looks wrong for multiple power modules.

File: src/modules/commander/esc_calibration.cpp [hunk 1]
```diff
@@ -105,14 +92,14 @@ int do_esc_calibration(orb_advert_t *mavlink_log_pub, struct actuator_armed_s* a
...
+	struct	battery_status_s battery = {};
 	int	batt_sub = -1;
 	bool	batt_updated = false;
-	bool	batt_connected = false;
+	bool	batt_connected = true;	// for safety resons assume battery is connected, will be cleared below if not the case
 
-	hrt_abstime battery_connect_wait_timeout = 30000000;
-	hrt_abstime pwm_high_timeout = 3000000;
-	hrt_abstime timeout_start;
+	hrt_abstime battery_connect_wait_timeout = 30*1000*1000;
```
Comment (dagar) [inline src/modules/commander/esc_calibration.cpp]: You can use @bkueng's new literals for this type of thing now. https://github.com/PX4/Firmware/blob/master/src/modules/land_detector/FixedwingLandDetector.h#L73

File: src/modules/commander/esc_calibration.cpp [hunk 2]
```diff
@@ -123,8 +110,15 @@ int do_esc_calibration(orb_advert_t *mavlink_log_pub, struct actuator_armed_s* a
 	}
 
 	// Make sure battery is disconnected
-	orb_copy(ORB_ID(battery_status), batt_sub, &battery);
-	if (battery.voltage_filtered_v > 3.0f) {
+	if (orb_copy(ORB_ID(battery_status), batt_sub, &battery) == PX4_OK) {
+
+		// battery is not connected if voltage is lower than 3V and we have a recent battery measurement
+		if (battery.voltage_filtered_v < 3.0f && (hrt_absolute_time() - battery.timestamp < 500*1000)) {
```
Comment (dagar) [inline src/modules/commander/esc_calibration.cpp]: You might want to start using hrt_elapsed_time() instead. There's been talk of adding additional checks within. https://github.com/PX4/Firmware/pull/8912
 
500*1000 -> 500_ms literal

Comment (dagar) [reply src/modules/commander/esc_calibration.cpp]: There's also a battery_status.connected field. Wouldn't that be better than checking a magic voltage?

File: src/modules/commander/esc_calibration.cpp [hunk 3]
```diff
@@ -105,14 +92,14 @@ int do_esc_calibration(orb_advert_t *mavlink_log_pub, struct actuator_armed_s* a
...
 	bool	batt_updated = false;
-	bool	batt_connected = false;
+	bool	batt_connected = true;	// for safety resons assume battery is connected, will be cleared below if not the case
 
-	hrt_abstime battery_connect_wait_timeout = 30000000;
-	hrt_abstime pwm_high_timeout = 3000000;
-	hrt_abstime timeout_start;
+	hrt_abstime battery_connect_wait_timeout = 30*1000*1000;
+	hrt_abstime pwm_high_timeout = 3*1000*1000;
+	hrt_abstime timeout_start = 0;
```
Comment (dagar) [inline src/modules/commander/esc_calibration.cpp]: Why do we need special behaviour for ocpoc here? Is actuator_armed.in_esc_calibration_mode not being correctly by commander? These type of scattered platform differences are horrible.

Comment (RomanBapst) [reply src/modules/commander/esc_calibration.cpp]: @dagar I was wondering as well when I looked at that code. Seems like it keeps the high pwm pulse for 60 seconds- seems like a very long time.

**Comment by RomanBapst** [comment]
@dagar I followed you suggestions, let's hope for fingers's sake that the 'connected' flag in the battery status is always set correctly and people remove their props before calibration.

**Comment by dagar** [comment]
We could do a quick check on some of the main targets.
 - px4fmu-v2
 - px4fmu-v3 (pixhawk cube 2.1) both power ports
 - px4fmu-v4
 - px4fmu-v4pro (dual power ports)
 - px4fmu-v5 (up to 4?)

There should be a battery_status publication for each brick, including a connected flag (set from ADC voltage > 1.5).

    listener battery_status

**Comment by dagar** [comment]
If this has been tested I'd like to merge it in time for the next release.

**Comment by r0gelion** [comment]
@dagar @RomanBapst could you give more details of the testing procedure for this PR, please?

**Comment by dagar** [comment]
@r0gelion ESC calibration with and without a battery (or power module) connected properly.

File: src/modules/commander/esc_calibration.cpp [hunk 4]
```diff
@@ -61,26 +61,15 @@
...
-	}
-	return PX4_OK;
-}
+using namespace time_literals;
 
 int do_esc_calibration(orb_advert_t *mavlink_log_pub, struct actuator_armed_s* armed)
 {
 	int	return_code = PX4_OK;
 	
 #if defined(__PX4_POSIX_OCPOC)
```
Comment (dagar) [inline src/modules/commander/esc_calibration.cpp]: Side note, this OCPOC specific code needs to be generalized at some point.

**Comment by santiago3dr** [comment]
ran initial test on a pixhawk mini with V2
WITH a power module connected, esc calibration works
BUT... the timeout is too fast (about 2~3 seconds) whereas esc calibration in master waits for a more reasonable period (20+ seconds)

WITHOUT a power module connected, esc calibration times out immediately upon connecting the battery
immidiate timeout

**Comment by RomanBapst** [comment]
@santiago3dr The esc calibration can only work if you have a power module connected. The reason is because the code uses the measured voltage from the power module to determine if the battery is connected or not.

**Comment by RomanBapst** [comment]
@santiago3dr You were right about the 3 second timeout, that was way too short. I set it from 3 to 20 seconds now.

**Comment by RomanBapst** [comment]
I tested this on a Pixhawk 4 and I have no issue with calibration. The timeout for connecting the battery is now set to 20 seconds which should give the user plenty of time.

**Pull request merged**